plutus-tx/src/PlutusTx/Blueprint/TH.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE CPP              #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE NamedFieldPuns   #-}
+{-# LANGUAGE RecordWildCards  #-}
+{-# LANGUAGE TemplateHaskell  #-}
+{-# LANGUAGE TupleSections    #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ViewPatterns     #-}
+
+module PlutusTx.Blueprint.TH where
+
+import Prelude
+
+import Data.List.NonEmpty qualified as NE
+import Data.Traversable (for)
+import Data.Typeable (Typeable)
+import GHC.Natural (Natural, naturalToInteger)
+import Language.Haskell.TH qualified as TH
+import Language.Haskell.TH.Datatype qualified as TH
+import PlutusTx.Blueprint.Class (HasDataSchema (..))
+import PlutusTx.Blueprint.Schema (HasSchemaDefinition, Schema (..))
+import PlutusTx.IsData.TH (makeIsDataIndexed)
+
+{- | Generate a 'ToData', 'FromData', 'UnsafeFromData', 'HasDataSchema' instances for a type,
+using an explicit mapping of constructor names to indices.
+Use this for types where you need to keep the representation stable.
+-}
+makeIsDataSchemaIndexed :: TH.Name -> [(TH.Name, Natural)] -> TH.Q [TH.InstanceDec]
```

Comment:
Some comments in this module would be helpful. TH can otherwise be quite hard to read.

And is there tests for this? If not then I'd again suggest that this be included for review only when it is used/tested.

Reply:
Yes, there is a Blueprint test that includes this functionality: `Blueprint.Tests` + `Blueprint.Tests.Lib`.

I'll add comments to explain TH functionality as you suggested.